core/dbt/config/project.py

```diff
@@ -244,7 +250,7 @@ class Project:
     snapshots: Dict[str, Any]
     dbt_version: List[VersionSpecifier]
     packages: Dict[str, Any]
-    query_comment: Optional[Union[str, NoValue]]
+    query_comment: Dict[str, Any]
```

Comment:
Instead of a `Dict[str, Any]`, let's make this a new optional `JsonSchemaMixin` `dataclass` with two fields:

```
@dataclass
class QueryComment(JsonsSchemaMixin):
    comment: Optional[str] = DEFAULT_QUERY_COMMENT
    append: bool = False
```

Then on the config/partial project objects, we'd change this to something like:
```
    query_comment: Optional[QueryComment] = field(default_factory=QueryComment)
```

We can use the object itself being `None` to indicate the old `None` behavior (don't include any comment) and the `comment` field being `None` to indicate the old `NoValue` behavior (default comment).

Reply:
Should we support the following config ( for backward compatibility) ?
```
query-comment: "executed by dbt"
```

The following definition will not allow string values for `query-comment`:
```
query_comment: Optional[QueryComment] = field(default_factory=QueryComment)
```

Reply:
Hmm, you're right - we'll want to make it a `Union[QueryComment, str]` and do an `isinstance` check at runtime!

Reply:
Do you have any idea why jsonschema validation fails for this config:
```
query-comment: 
  comment: "executed by dbt"
  append: true
```


core/dbt/contracts/project.py
```python
@dataclass
class Project(HyphenatedJsonSchemaMixin, Replaceable):
    ...
    query_comment: Optional[Union[QueryComment, str]] = None
```

```python
@dataclass
class QueryComment(JsonSchemaMixin):
    comment: str = DEFAULT_QUERY_COMMENT
    append: bool = False
```

Here is the error message :
```
2020-03-17 21:04:42.892154 (MainThread): Running with dbt=0.17.0-a1
2020-03-17 21:04:43.124426 (MainThread): Encountered an error while reading the project:
2020-03-17 21:04:43.124548 (MainThread):   ERROR: Runtime Error
  at path []: QueryComment(comment='executed by dbt', append=True) is not of type 'object'
2020-03-17 21:04:43.124647 (MainThread): Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'invalid', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7fd85acdb370>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7fd85acdb940>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x7fd85acdb2e0>]}
2020-03-17 21:04:43.124892 (MainThread): Encountered an error:
2020-03-17 21:04:43.124948 (MainThread): Runtime Error
  Could not run dbt
2020-03-17 21:04:43.126486 (MainThread): jsonschema.exceptions.ValidationError: QueryComment(comment='executed by dbt', append=True) is not of type 'object'

Failed validating 'type' in schema[0]:
    {'additionalProperties': False,
     'description': 'QueryComment(comment: str = "\\n{%- set comment_dict '
                    '= {} -%}\\n{%- do comment_dict.update(\\n    '
                    "app='dbt',\\n    dbt_version=dbt_version,\\n    "
                    "profile_name=target.get('profile_name'),\\n    "
                    "target_name=target.get('target_name'),\\n) -%}\\n{%- "
                    'if node is not none -%}\\n  {%- do '
                    'comment_dict.update(\\n    '
                    'node_id=node.unique_id,\\n  ) -%}\\n{% else %}\\n  {# '
                    'in the node context, the connection name is the '
                    'node_id #}\\n  {%- do '
                    'comment_dict.update(connection_name=connection_name) '
                    '-%}\\n{%- endif -%}\\n{{ return(tojson(comment_dict)) '
                    '}}\\n", append: bool = False)',
     'properties': {'append': {'default': False, 'type': 'boolean'},
                    'comment': {'default': '\n'
                                           '{%- set comment_dict = {} -%}\n'
                                           '{%- do comment_dict.update(\n'
                                           "    app='dbt',\n"
                                           '    dbt_version=dbt_version,\n'
                                           '    '
                                           "profile_name=target.get('profile_name'),\n"
                                           '    '
                                           "target_name=target.get('target_name'),\n"
                                           ') -%}\n'
                                           '{%- if node is not none -%}\n'
                                           '  {%- do comment_dict.update(\n'
                                           '    node_id=node.unique_id,\n'
                                           '  ) -%}\n'
                                           '{% else %}\n'
                                           '  {# in the node context, the '
                                           'connection name is the node_id '
                                           '#}\n'
                                           '  {%- do '
                                           'comment_dict.update(connection_name=connection_name) '
                                           '-%}\n'
                                           '{%- endif -%}\n'
                                           '{{ '
                                           'return(tojson(comment_dict)) '
                                           '}}\n',
                                'type': 'string'}},
     'required': [],
     'type': 'object'}

On instance:
    QueryComment(comment='executed by dbt', append=True)
```

I'd prefer asking  before  digging into details of jsonschema .

Thank you

Reply:
I usually see that when I put the actual object in a place where we expected the dict form.